docs/proposals/24-shoot-oidc-issuer.md

```diff
@@ -0,0 +1,130 @@
+ - Provide a secure managed solution for serving the OIDC discovery documents of a shoot cluster.
+ - Improve integration capabilities for clusters that reside in a fenced environment.
+### Non-Goals
+ - Establish trust relations between clusters. This GEP is about simplifying the prerequisites for doing that.
+
+## Proposal
+
+This proposal consists of three parts that together will achieve the mentioned goals.
+
+### Static validation
+
+Static validation rule will be introduced enforcing that `shoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer` is not set when the shoot is annotated with `gardener.cloud/managed-issuer: true`. This annotation will be incompatible with workerless shoots, as they do not have nodes and they do not run any workloads. The rule will also assure that once present the annotation cannot be removed from a shoot.
+
+```go
+if oldShoot.Annotations["gardener.cloud/managed-issuer"] == "true" {
+    // ensure that the new shoot also has this annotation
+    // ensure that newShoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer is not set
+    // ensure that the shoot is not configured as workerless
+    newShoot.Annotations["gardener.cloud/managed-issuer"] = "true"
+} else if newShoot.Annotations["gardener.cloud/managed-issuer"] == "true" {
+    // ensure that newShoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer is not set
+    // ensure that the shoot is not configured as workerless
+}
+```
+
+### Gardenlet
+
+Once annotated shoots will be easily identifiable by the `gardenlet`. If a shoot requires a managed issuer then the `gardenlet` will configure the `--service-account-issuer` and `--service-account-jwks-uri` flags for the shoot's [kube-apiserver](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) deployment during reconciliation and set their values to `https://<central.domain.name>/projects/<project-name>/shoots/<shoot-uid>/issuer` and `https://<central.domain.name>/projects/<project-name>/shoots/<shoot-uid>/issuer/jwks` respectively (the central domain name should be configurable via the gardenlet's configuration). At the end of a shoot reconciliation flow, the `gardenlet` will fetch the OIDC metadata discovery documents from a shoot's `kube-apiserver` and sync them back to the Garden cluster in the form of a configmap following the naming convention `projectname--shootuid` in a special purposed namespace called `gardener-shoot-issuer`(name should be configurable). The [Seed Authorizer](https://github.com/gardener/gardener/blob/master/docs/deployment/gardenlet_api_access.md#scoped-api-access-for-gardenlets-and-extensions) will be enhanced so that Gardenlet is only permitted to perform actions on related configmaps. An additional advertised address will be added to the shoot status so that the issuer of the cluster is easily discoverable.
```

Comment:
> At the end of a shoot reconciliation flow

It could do it right after it waited for KAPI to be rolled out, or?

Reply:
Yes, I think this could happen immediately after KAPI rollout. I will update the proposal with this change

Reply:
Oh, I missed that. Why don't we deploy this `ConfigMap into the namespace of the shoot? This is also where we put other shoot-related information like CA bundle or observability endpoints.

Reply:
There are several reasons that I do not suggest copying the configmap directly in the shoot namespace.

1. The shoot metadata server will need to have access to configmaps in all namespaces instead of one
2. The configmaps could be modified by users. Even if they are immutable, they can be deleted and recreated meaning that users will be able to control what the metadata server serves. This is particularly undesirable from my point of view since a user with malicious intents and access to the configmaps can for example self generate a signing key, then put the corresponding public key in the configmap and sign tokens that will be forwarded to external systems. And these external systems will think that these are actually tokens signed by the `kube-apiserver` and will give them access, when in fact these tokens are actually signed by a malicious actor.